numpy/random/mtrand/randint_helpers.pxi.in

```diff
@@ -51,27 +57,61 @@ def _rand_{{npy_dt}}(low, high, size, rngstate):
     -------
     out : python integer or ndarray of np.{{np_dt}}
           `size`-shaped array of random integers from the appropriate
-          distribution, or a single such random int if `size` not provided.
+          distribution, or an array corresponding to the shapes of
+          `low` and `high`.
 
     """
-    cdef npy_{{npy_udt}} off, rng, buf
-    cdef npy_{{npy_udt}} *out
-    cdef ndarray array "arrayObject"
-    cdef npy_intp cnt
     cdef rk_state *state = <rk_state *>PyCapsule_GetPointer(rngstate, NULL)
+    cdef ndarray array "arrayObject"
+    cdef npy_{{npy_udt}} off, rng, buf
+    cdef npy_{{npy_udt}} *array_data
+
+    cdef ndarray lo, hi
+    cdef npy_intp i, cnt
+    cdef broadcast multi
+
+    lo = <ndarray>np.array(low)
+    hi = <ndarray>np.array(high)
 
-    rng = <npy_{{npy_udt}}>(high - low)
-    off = <npy_{{npy_udt}}>(<npy_{{npy_dt}}>low)
+    if lo.shape == hi.shape == ():
+        low = int(low)
+        high = int(high)
 
-    if size is None:
-        rk_random_{{npy_udt}}(off, rng, 1, &buf, state)
-        return np.{{np_dt}}(<npy_{{npy_dt}}>buf)
+        rng = <npy_{{npy_udt}}>(high - low)
+        off = <npy_{{npy_udt}}>(<npy_{{npy_dt}}>low)
+
+        if size is None:
+            rk_random_{{npy_udt}}(off, rng, 1, &buf, state)
+            return np.{{np_dt}}(<npy_{{npy_dt}}>buf)
+        else:
+            array = <ndarray>np.empty(size, np.{{np_dt}})
+            cnt = PyArray_SIZE(array)
+            array_data = <npy_{{npy_udt}} *>PyArray_DATA(array)
+            with nogil:
+                rk_random_{{npy_udt}}(off, rng, cnt, array_data, state)
+            return array
     else:
-        array = <ndarray>np.empty(size, np.{{np_dt}})
-        cnt = PyArray_SIZE(array)
+        i = 0
+
+        if size is None:
+            multi = <broadcast>np.broadcast(lo, hi, [0])
+            array = <ndarray>np.empty(multi.shape, dtype=np.{{np_dt}})
+        else:
+            array = <ndarray>np.empty(size, dtype=np.{{np_dt}})
+            multi = <broadcast>np.broadcast(lo, hi, array)
+            if multi.shape != array.shape:
+                raise ValueError("size is not compatible with inputs")
+
         array_data = <npy_{{npy_udt}} *>PyArray_DATA(array)
-        with nogil:
-            rk_random_{{npy_udt}}(off, rng, cnt, array_data, state)
+
+        for low_data, high_data, array_elt in multi:
+            rng = <npy_{{npy_udt}}>(int(high_data) - int(low_data))
+            off = <npy_{{npy_udt}}>(<npy_{{npy_dt}}>low_data)
+
+            rk_random_{{npy_udt}}(off, rng, 1, &buf, state)
```

Comment:
nit: Can we take these three lines and put them in a helper `cdef rk_random_{{npy_udt}}_bounds`? Would be nice to not have fiddly casting duplicated